solved stam exercise

kerenGuez opened the pull request.

kerenGuez committed:
solved stam exercise

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,43 @@
+__author__ = "Keren"
```
No need

rfire01 replied:
The first line in the file should be a description of the purpose of the file

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def add_vals(line):
+    """
```
The first line of docs should be a description of the function \ class , etc

rfire01 replied:
look at chapter 3.8 how the docs should look like
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,43 @@
+def add_vals(line):
+    """
+    Args:
+        line: (str) a line of text
```
the type should be before the ":"

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,43 @@
+
+
+def stam(filename):
+    """ The function gets file path, edits it and saves the new edited file to a new location
```
look at pep8 conventions (how many chars per line, space line after logic block, etc.)
https://realpython.com/python-pep8/#blank-lines

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,43 @@
+        filename: (str) the full path to the file that will be edited
+
+    """
+    try:
```
you can instead check if file exists at the start

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,43 @@
+    try:
+        with open(filename, 'r') as file:
+            new_concat_lines_lis = (add_vals(line[:-1])+"\n" if line[-1:] == "\n" else add_vals(line) for line in file)
+            new_path_to_save = r"C:\Users\keren\Desktop\newfile.txt"
```
you should put the output path as function argument

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,43 @@
+            new_path_to_save = r"C:\Users\keren\Desktop\newfile.txt"
+            with open(new_path_to_save, 'w') as new_file:
+                for line in new_concat_lines_lis:
+                    print("line "+line, end="")
```
no need

kerenGuez committed:
-I fixed the problematic Docstrings
-I changed the try and except to a condition that checks if file paths exist
-I added the output file path as an argument to the function
-I removed unnecessary print line

kerenGuez committed:
small change to a docstring

kerenGuez committed:
shortened a line

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,49 @@
+"""This program reads a file, edits it and saves it in a new location"""
+import os
+FIRST_NAME = "Keren"
+LAST_NAME = "Guez"
```
imports and the code should be separated with blank lines

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def add_vals(line):
+    """This function concatenates a line with some given text
```
dot at end of sentence (in all docs)

rfire01 replied:
functions doc should start with Imperative verb.
For example: Concatenate a line with...
Add something with
etc.

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,49 @@
+
+    Returns:
+        str: a new concatenated line
+
```
remove

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,49 @@
+
+    """
+    last_slash = output_file.rfind("\\")
+    if os.path.exists(filename) and os.path.exists(output_file[:last_slash]):
```
you don't need to check if output file exists, as you are creating it now

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,49 @@
+                for line in new_concat_lines_lis:
+                    new_file.write(line)
+    else:
+        print(f"either the original file path or the output path don't exist")
```
error messages should be informative.
Don't let the user guess what the problem, say exactly, is it the orignal or output?

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,49 @@
+            new_concat_lines_lis = (add_vals(line[:-1])+"\n"
+                                    if line[-1:] == "\n"
+                                    else add_vals(line)
+                                    for line in file)
```
this doesn't look good, try to think on a better way

kerenGuez committed:
fixed the mistakes

kerenGuez committed:
solved atm exercise

kerenGuez committed:
solved

kerenGuez committed:
Delete AtmExAnswer.py

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,51 @@
+
+    Args:
+        input_file (str): The path to the file that will be edited.
+        output_folder (str) : The path to the output folder.
```
blank line between args and returns

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def check_exists(input_file, output_folder):
+    """Checks if paths are valid and prints a message accordingly.
```
Like I said imperative verbs (Check not checks), in both functions

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,51 @@
+        output_folder (str) : The path to the output folder.
+    Returns:
+        True : If both paths are valid.
+        False : If any path isn't valid.
```
in returns you write the type of the returned value and what it represents, not its values

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,51 @@
+    if not os.path.exists(output_folder):
+        print(f"The output folder '{output_folder}' doesn't exist.")
+        return False
+    return True
```
pep8

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,51 @@
+        output_file (str) : The path of the new file to be saved.
+    """
+    output_dir = os.path.dirname(os.path.abspath(output_file))
+    if check_exists(filename, output_dir):
```
instead of putting all the logic under if, you can first check it, if its not exists exits the function, and if not do the logic

osherdp commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,51 @@
+        True : If both paths are valid.
+        False : If any path isn't valid.
+    """
+    if not os.path.exists(input_file):
```
the `exists` usage is redundant
If one of the files doesn't exist it will raise an exception nevertheless

osherdp commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,51 @@
+    output_dir = os.path.dirname(os.path.abspath(output_file))
+    if check_exists(filename, output_dir):
+        with open(filename, 'r') as file:
+            new_concat_lines = (FIRST_NAME + line[:-1] + LAST_NAME + "\n"
```
move into the `for` below. For example:
```
for line in file:
    new_file.write(FIRST_NAME + line[:-1] + ...)
```

osherdp replied:
don't use string concatenation, use f-strings

osherdp reviewed (changes requested):
It's a very elegant solution, just a few fixes are needed

kerenGuez committed:
stam fixed

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,35 @@
+INPUT_FILE = 1
+OUTPUT_FILE = 2
+FIRST_NAME = "Keren"
+LAST_NAME = "Guez"
```
reorder by length

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,35 @@
+        new_path_to_save = sys.argv[OUTPUT_FILE]
+        stam(path, new_path_to_save)
+    except Exception as e:
+        print(e)
```
The try except here are pointless.
At the moment, you catch the exception and just print it as it is and ending the program. If this the case, you can just remove the try except, and let it throw exception if it fails.
If you do want to use try except here, you should do something different, such as printing your own error message

rfire01 commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,35 @@
+    try:
+        path = sys.argv[INPUT_FILE]
+        new_path_to_save = sys.argv[OUTPUT_FILE]
+        stam(path, new_path_to_save)
```
blank line after logic block

osherdp commented on basic/stam/StamExAnswer.py:
```diff
@@ -0,0 +1,35 @@
+"""Reads a file, edits it and saves it in a new location."""
```
Please make sure you name all python files like so:
* No capital letters, only small letters, underscores and (rarely) numbers
* Most meaningful name possible. In this training, you can remove usage of words like "exercise" and "answer", as it's not the essence of the code written
* Short as possible, but not too short if readability will suffer from it

kerenGuez committed:
fixed latest problems- renamed, added new line after logic block, added specific exception messages

kerenGuez committed:
fixed

rfire01 commented on basic/stam/stam.py:
```diff
@@ -0,0 +1,40 @@
+FIRST_NAME = "Keren"
+LAST_NAME = "Guez"
+OUTPUT_FILE = 2
+INPUT_FILE = 1
```
reorder

rfire01 commented on basic/stam/stam.py:
```diff
@@ -0,0 +1,40 @@
+    except IndexError:
+        print("Wrong numbers of parameters in script(should be 2).")
+    except FileNotFoundError as e:
+        print(f"The path '{e.filename}' was not found")
```
pep8

rfire01 commented on basic/stam/stam.py:
```diff
@@ -0,0 +1,40 @@
+    except FileNotFoundError as e:
+        print(f"The path '{e.filename}' was not found")
+    except Exception as e:
+        print(e)
```
It is ok for your code to throw an exception.
If you just catch it and print it to the screen it doesn't give additional information.

rfire01 commented on basic/stam/stam.py:
```diff
@@ -0,0 +1,40 @@
+
+
+if __name__ == "__main__":
+    main()
```
blank line at the end of file (pep8)

rfire01 replied:
?

kerenGuez committed:
Added spaces and reordered

kerenGuez committed:
blank line(for some reason it didn't show before)

kerenGuez committed:
.
